working/1847 - FinalizationRegistry/proposal.md

```diff
@@ -0,0 +1,379 @@
+side-effects observable from the pure Dart code.
+
+### Isolate-independent native functions
+
+`dart:ffi` allows developers to construct a native function pointer to a
+static function defined in Dart via [`Pointer.fromFunction`] constructor.
+This constructor essentially returns a pointer to a _trampoline_ which
+follows C ABI and can be called from native code. When invoked such trampoline
+will perform transition from native code into Dart code, marshal arguments,
+call the target static function, marshaling the result it returns and then
+transition back into native code. There is an implicit expectation baked into
+this process: calling a native-to-Dart trampoline will only succeed if
+there is a Dart isolate associated with the thread on which we perform the call
+and this isolate is in the state which allows reentrancy.
+
+Such isolate-dependent function can't be used as a finalization callback because
+finalization callbacks should be callable in contexts when there is no current
+isolate at all or isolates are not allowing entering into Dart code.
+
+It seems thus reasonable to restrict `NativeFinalizationRegistry` constructor
+in a way that would reject function pointers which are pointing to
+native-to-Dart FFI trampolines.
+
+This restriction however means that users might be required to write native code
+to implement their finalizers. Consider for example [`mmap`/`munmap`][mmap API]
+POSIX APIs.
+
+```cpp
+// mmap -- allocate memory, or map files or devices into memory
+void *mmap(void *addr, size_t len, int prot, int flags, int fd, off_t offset);
+
+// munmap -- remove a mapping
+int munmap(void *addr, size_t len);
+```
+
+A Dart developer will be able to call `mmap` and `munmap` through FFI, but they
+will not be able to use `munmap` directly as a finalizer, because it does not
+conform to `Void Function(Pointer<Void>)` signature and expects the size of
+the mapping as well. A developer would have to implement a helper in a native
+language instead:
+
+```cpp
+// C++
+struct Mapping {
+  void* addr;
+  uint64_t size;
+};
+
+void FinalizeMapping(Mapping* mapping) {
+  munmap(mapping->addr, mapping->size);
+  free(mapping);
+}
+```
+
+```dart
+// Dart
+
+final NativeFinalizerPtr finalizeMapping = lib.lookup('FinalizeMapping');
+final mmap = DynamicLibrary.process().lookupFunction<...>('mmap');
+
+class _Mapping extends Struct {
+  external Pointer<Void> addr;
+
+  @Uint64()
+  external int size;
+}
+
+class Mapping extends Finalizable {
+  final Pointer<_Mapping> mapping;
+
+  factory Mapping(int len) {
+    final mapping = malloc.alloc<_Mapping>(sizeOf<_Mapping>());
+    mapping.addr = mmap(len, ...);
+    mapping.len = len;
+    final wrapper = Mapping._(mapping);
+    _registry.register(wrapper, mapping, externalSize: len);
+    return wrapper;
+  }
+
+  static final _registry = NativeFinalizationRegistry(finalizeMapping);
+}
+```
+
+It would be more convenient if Dart developer did not need to write any native
```

Comment:
For packages targeting system APIs (for example `package:win32`), they decide to not write any native code at all. This would mean that for those packages finalizers are useless unless we actually implement the idea below.

Reply:
Ack.